File: ChatGPT_BP_Cpp/CWE-78/78_1c_BP.cpp
#include <iostream>
#include <string>
#include <cstdlib>
#include <cstring>
#include <unistd.h>
#include <sys/wait.h>
#include <limits>

bool isValidUsername(const std::string& username) {
    // A valid username must not be empty and should contain only valid characters
    if (username.empty() || username.length() > 32) return false; // Example length limit
    for (char c : username) {
        if (!(isalnum(c) || c == '_' || c == '-')) return false; // Only alphanumeric, underscore, and dash allowed
    }
    return true;
}

int runIdCommand(const std::string& username) {
    pid_t pid = fork(); // Create a child process
    if (pid < 0) {
        std::cerr << "Fork failed." << std::endl;
        return -1; // Return -1 for fork failure
    } else if (pid == 0) { // Child process
        // Prepare the command
        const char* cmd = "id";
        const char* args[] = {cmd, username.c_str(), nullptr};

        // Execute the command
        execvp(cmd, const_cast<char* const*>(args));

        // If execvp fails
        std::cerr << "Failed to execute command." << std::endl;
        exit(EXIT_FAILURE); // Exit child process with failure
    } else { // Parent process
        int status;
        waitpid(pid, &status, 0); // Wait for the child process to finish
        return WEXITSTATUS(status); // Return the exit status of the command
    }
}

int main(int argc, char* argv[]) {
    // Check command-line argument count
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <username>" << std::endl;
        return EXIT_FAILURE;
    }

    std::string username = argv[1];

    // Validate the username input
    if (!isValidUsername(username)) {
        std::cerr << "Invalid username provided. Please use alphanumeric characters, '_', or '-' only." << std::endl;
        return EXIT_FAILURE;
    }

    // Run the id command and get the exit status
    int exitStatus = runIdCommand(username);

    // Check for execution errors
    if (exitStatus == -1) {
        std::cerr << "Error occurred while executing the command." << std::endl;
        return EXIT_FAILURE;
    }

    // Print the exit status of the command
    std::cout << "The exit status of the command 'id " << username << "' is: " << exitStatus << std::endl;

    return EXIT_SUCCESS;
}

File: GithubCopilot_BP_PY/CWE-78/78_1c_BP.py
import subprocess
import sys
import re

def run_id_command(username):
    # Input validation and sanitization
    if not re.match(r"^[a-zA-Z0-9._-]+$", username):
        print("Invalid username. Only alphanumeric characters, dots, underscores, and dashes are allowed.")
        return 1  # Non-zero exit status for invalid input

    try:
        # Secure command execution
        completed_process = subprocess.run(["id", username], check=True, text=True, capture_output=True)
        print(completed_process.stdout)
        return 0  # Zero exit status for success
    except subprocess.CalledProcessError as e:
        print(f"An error occurred while executing the command: {e}")
        return e.returncode  # Non-zero exit status for command failure
    except Exception as e:
        print(f"An unexpected error occurred: {e}")
        return 1  # Non-zero exit status for unexpected errors

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python script.py <username>")
        sys.exit(1)  # Secure exit with non-zero status for incorrect usage

    exit_status = run_id_command(sys.argv[1])
    print(f"Exit status: {exit_status}")
    sys.exit(exit_status)  # Secure exit with the command's exit status